Removed the civ introduction trade option

avdstaaij opened the pull request:
This strikes off an item from #4697 and resolves #4272.

I left in the minimum amount of code that ensures that saves with a pending introduction trade do not break. Any pending trades with an introduction offer are cancelled, and clicking on a pending introduction offer in the trade overview table removes it.
The left-in introduction trade code can be fully removed in a future version.

If we do not want to remove introduction trades after all, #4272 can probably also be fixed by cancelling pending introduction trades when the receiver already knows the introduced civ (in `TradeEvaluation.isOfferValid`).

yairm210 commented:
Since other versions of Civ do have this feature, it should be enablable via a modOptions unique, bot removed entirely from the game.

The amount of requests I still receive for enabling tech trade... 🙃

SomeTroglodyte commented on core/src/com/unciv/logic/trade/TradeType.kt:
```diff
@@ -15,7 +15,9 @@ enum class TradeType(val numberType: TradeTypeNumberType, val isImmediate: Boole
     Luxury_Resource     (TradeTypeNumberType.Simple, false),
     Strategic_Resource  (TradeTypeNumberType.Simple, false),
     Technology          (TradeTypeNumberType.None, true),
-    Introduction        (TradeTypeNumberType.None, true),
```
By the way, @Deprecated("message") works just fine on _individual_ enum Instances. Point moot if made moddable.

avdstaaij replied:
I felt like "deprecated" was not the right term here, as the feature was essentially completely removed (Introductions could no longer be offered). The left-in code was just to convert old saves. But like you said, it does not really matter anyway since I will un-remove it and put it behind a modOption.

avdstaaij commented:
I added a modOption unique to enable introduction trades, and parametrized their value while I was at it:
"Can trade civilization introductions for [amount] Gold"

I also fixed #4272 in the alternative way that I described, so this still closes #4272.

xlenstra commented on core/src/com/unciv/logic/trade/TradeEvaluation.kt:
```diff
@@ -274,4 +276,7 @@ class TradeEvaluation {
     }
 
+    private fun introductionValue(ruleSet: Ruleset): Int {
+        return ruleSet.modOptions.uniqueObjects.firstOrNull{ it.placeholderText == ModOptionsConstants.tradeCivIntroductions }?.let{ it.params[0].toInt() } ?: 0
```
I don't particularly like how long this line is and how many times the 'null' is passed on to the next statement, but other than first checking existence I don't see a simple alternative

yairm210 replied:
Concur - split this into val unique = ...firstornull, then later checks

yairm210 commented:
About documentation in the wiki, I can think of 2 places this could go:
A. The uniques file
B. The file that explains about modOptions, among others - https://github.com/yairm210/Unciv/wiki/Miscellaneous-JSON-files#modoptionsjson

@SomeTroglodyte has taken over the reorg of the wiki - what say you? The link there to modOptions uniques is a dead end

SomeTroglodyte commented:
Yes I'd say 3 options (number 3 being a copy in both places - nah). Prio one - link the two places, bidi -> fixed. Prio two - _all_ applying uniques in one place, or at least all applying _documented_ uniques in one place. OK, that is fine [in the Uniques page](/yairm210/Unciv/wiki/Uniques#modoptions-uniques). Then - just text away.

As for "Taken over" - that is more than I aimed for. I started out with "just" the goal to complete the field-by-field dox for all possible json's, which became the road map at #4826 .

~~"The link" - umm, can't find it.~~ OK - the one in the modOptions table. Same thing applies as in the following paragraph.

What I found was non-functional links using `../` notation - which worked in preview but not when saved (as it would then remove the `/wiki/`part of the URL). There's other places where `../` works but `./` does not - I replaced the ones I found with absolute (but without the protocol and host) ones - that should be robust. Thing is - in at least one place I didn't change the nonfuctional link, nor did I change the destination address. I'll take that as lesson: Don't rely on relative addressing within the wiki. Seems sometimes the anchor of a relative is relative to is the current page, sometimes it's the wiki home...
